### ISSUE TITLE
`sort_keys=True` simplejson encoder option can result in incorrect object references

### ISSUE THREAD

**Issue by mspacek** [jsonpickle/jsonpickle#408]
I like to use the `sort_keys=True` simplejson encoder option so that attributes of my serialized object(s) are in alphabetical order, which makes them more human readable. However, in jsonpickle > 1.3 (including the latest 2.2 release), I've noticed that reliably causes problems if one or more of the attributes are themselves more complex objects that are referenced at least once, and if the attributes are not originally defined in alphabetical order. It's as if first jsonpickle calculates references, then simplejson sorts the underlying objects, and then during decoding, jsonpickle uses the original references on the sorted objects, resulting in reference errors.

Here's a minimal working example:

```python
"""When sort_keys=True in simplejson, jsonpickle > 1.3 dereferences the frozen attribs
incorrectly, and a3 incorrectly references a2 instead of a1"""

sort_keys = True
keys = True

import sys
import simplejson
import jsonpickle
jsonpickle.set_preferred_backend('simplejson')
jsonpickle.set_encoder_options('simplejson', indent=' ', separators=(',', ':'),
                               sort_keys=sort_keys)
from hashlib import md5

class A(object):
    def __init__(self):
        self.id = md5(str(id(self)).encode()).hexdigest()[:5] # unique enough hash

class B(object):
    def __init__(self):
        self.a2 = A() # set attribs not in alphabetical order
        self.a1 = A()
        self.a3 = self.a1 # create a reference

b = B()
print('Python %s' % sys.version)
print('simplejson.__version__ = %s' % simplejson.__version__)
print('jsonpickle.__version__ = %s' % jsonpickle.__version__)
print('sort_keys = %s' % sort_keys)
print('b.a1.id = %s, b.a2.id = %s, b.a3.id = %s' % (b.a1.id, b.a2.id, b.a3.id))

frozen = jsonpickle.encode(b, keys=keys, warn=True)
print(frozen)

thawed = jsonpickle.decode(frozen, keys=keys)
print('b.a1.id = %s, b.a2.id = %s, b.a3.id = %s' % (thawed.a1.id, thawed.a2.id, thawed.a3.id))
```

Here's the output with `sort_keys=True`:

```
Python 3.8.10 (default, Jun 22 2022, 20:18:18) 
[GCC 9.4.0]
simplejson.__version__ = 3.17.6
jsonpickle.__version__ = 2.2.0
sort_keys = True
b.a1.id = 45d5e, b.a2.id = 26176, b.a3.id = 45d5e
{
 "a1":{
  "id":"45d5e",
  "py/object":"__main__.A"
 },
 "a2":{
  "id":"26176",
  "py/object":"__main__.A"
 },
 "a3":{
  "py/id":2
 },
 "py/object":"__main__.B"
}
b.a1.id = 45d5e, b.a2.id = 26176, b.a3.id = 26176
```

Notice that `b.a3` incorrectly references `b.a2` instead of `b.a1`. This is resolved by setting `sort_keys=False`:

```
Python 3.8.10 (default, Jun 22 2022, 20:18:18) 
[GCC 9.4.0]
simplejson.__version__ = 3.17.6
jsonpickle.__version__ = 2.2.0
sort_keys = False
b.a1.id = 20db0, b.a2.id = bc504, b.a3.id = 20db0
{
 "py/object":"__main__.B",
 "a2":{
  "py/object":"__main__.A",
  "id":"bc504"
 },
 "a1":{
  "py/object":"__main__.A",
  "id":"20db0"
 },
 "a3":{
  "py/id":2
 }
}
b.a1.id = 20db0, b.a2.id = bc504, b.a3.id = 20db0
```

The referencing problem also goes away if I originally assign the attribs in alphabetical order. The `keys` encode and decode kwarg doesn't have an effect.

Apologies if this is already known, but it took me a long time to figure this out, doesn't seem to be documented, and results in mostly silent round-trip data corruption, so I'm pretty sure this is a bug.

Is there any way currently to avoid this referencing error while still using `sort_keys=True` in simplejson and `make_refs=True` in `jsonpickle.encode()`?

**Comment by Theelx** [comment]
Hey, thanks so much for reporting this! You've made a very well-reported and well-researched issue, so I appreciate that. This doesn't seem to be previously known.

I'll take a look into this, but the ID/ref system is not something I fully understand, so it could take a while to resolve.

**Comment by Theelx** [comment]
Update: I solved this for objects using `__dict__` in about 5 minutes, which was surprisingly easy. However, classes with `__slots__` will be extremely difficult to fix (without subclassing, which could cause extra issues) as that's read-only and can't be sorted. I'm going to see what I can do over the next few days, and if I can't fix `__slots__` then I'll just release handling only for classes with `__dict__` and raise a warning if the user wants classes with `__slots__` to be sorted.

**Comment by Theelx** [comment]
I'm going to just make it raise a warning with `__slots__`, as it seems very difficult to fix that case and I want to try and get 3.0.0 out soon.

**Comment by Theelx** [comment]
Fixed in https://github.com/jsonpickle/jsonpickle/commit/71fa3b55354d078c1dc383d2a8e35307bc34c855
